### PR TITLE
Implemented advanced text functions

### LINKED ISSUE #20: Ghost keyboard also for selections
Would be very nice if the ghost keyboard (long pressed space bar) could appear also when a text is select so moving the limits of the selection would be more precisely.
Maybe the keyboard could be split in two: left for first cursor and right for the second one.
Another thing I thought is the ghost keyboard could remain until I press the space key, because selecting text could take more time imo...
hoping I explained enough well... ;)

### PR DESCRIPTION
**Pull request by kugiigi**
### Implementations
 * Added a new toolbar with functions such as copy, cut, and paste (available in cursor swipe mode)
 * Added a way to select texts in the cursor (double-tapping while in cursor swipe mode)
 * Added floating buttons to exit cursor swipe mode and place the cursor in the start/end of line/document  (available in cursor swipe mode)

### Changes/Fixes
 * Added audio and haptic feedback to the space key and word ribbon
 * Changed cursor swipe mode timeout from 400ms to 5000ms. This is to give users enough time for the new toolbar. The done button gives users a way to immediately exit the cursor swipe mode.
 * Makes background of the cursor swipe mode fully solid instead of `0.5` opacity.
 * Fixes issues with the cursor swipe mode in QtWebEngine text fields. This was done by commenting out `input_method.surroundingRight` and `input_method.surroundingLeft` from the criteria for moving the cursor. This however has an effect which now allows moving the cursor between lines with horizontal swipes. This needs to be revisited once the issue with QtWebEngine gets fixed.
 * Moved audio and haptic feedback components from `KeyboardComponent.qml` to `Keyboard.qml` and created a function that can be used to enable both feedback easily.

### Screenshots
Ambiance Theme:

**Cursor Swipe Mode**
screenshot20190617_174058792

**Selection Mode**
screenshot20190617_174105031

Fixes https://github.com/ubports/keyboard-component/issues/20

### CODE REVIEW

File: qml/Keyboard.qml
Comment (UniversalSuperBox) [inline qml/Keyboard.qml:51]: Is there any reason to have this time out now that there is a "Done" button?

File: qml/Keyboard.qml
Comment (UniversalSuperBox) [inline qml/Keyboard.qml:388]: Do you have a bug reference for this comment?

File: qml/Keyboard.qml
Comment (UniversalSuperBox) [inline qml/Keyboard.qml:583]: Also needs a bug report attached

File: src/view/abstracttexteditor.cpp
Comment (UniversalSuperBox) [inline src/view/abstracttexteditor.cpp]: This should either be left in with a way to set the debug attribute or left out.

File: src/view/abstracttexteditor.cpp
Comment (UniversalSuperBox) [inline src/view/abstracttexteditor.cpp]: It's not clear what this means to me. If you think it's needed to have a comment, be very specific with what you mean

File: qml/Keyboard.qml
Comment (UniversalSuperBox) [inline qml/Keyboard.qml:337]: We shouldn't distinguish the mode only by the color of the keyboard background. This helps people who aren't able to distinguish between blue and their primary keyboard color.

Should we add some text to the keyboard, such as "Drag here to move cursor" or "Drag here to move selection"?

File: qml/Keyboard.qml
Comment (UniversalSuperBox) [inline qml/Keyboard.qml:603]: I found it was a little difficult to go up or down if a line was blank. The cursor jumps around quite a bit, probably because it's sending left or right keypresses at the same time as the up or down.

**Review by UniversalSuperBox — CHANGES_REQUESTED** [review]
Wow, that's quite a change. There's a lot going on here, so I think that reviews will find different points as they continue. Remember that it's not because we want to keep beating on you, but because we keep finding more stuff that we can fix to make the best keyboard around! :)

File: qml/Keyboard.qml
Comment (kugiigi) [reply qml/Keyboard.qml:51]: I decided to still have a timeout albeit quite longer than before. Honestly, I'm not sure if it still make sense or not but I think it's fine leaving it there.

File: qml/Keyboard.qml
Comment (kugiigi) [reply qml/Keyboard.qml:388]: Here's the bug about it although I was also the one that raised it and it's still tag as "needs confirmation"
https://github.com/ubports/ubuntu-touch/issues/1157

File: qml/Keyboard.qml
Comment (kugiigi) [reply qml/Keyboard.qml:583]: https://github.com/ubports/ubuntu-touch/issues/1157

File: src/view/abstracttexteditor.cpp
Comment (kugiigi) [reply src/view/abstracttexteditor.cpp]: Yep, I think I should delete that comment.....I might be a victim of being a "Copy-paste programmer" :grinning:

File: src/view/abstracttexteditor.cpp
Comment (kugiigi) [reply src/view/abstracttexteditor.cpp]: Hmmm...I myself don't fully understand it...probably one of the things in my old branch that isn't applicable anymore. I will remove it.

File: qml/Keyboard.qml
Comment (kugiigi) [reply qml/Keyboard.qml:337]: I will try to add some texts for both modes.

File: qml/Keyboard.qml
Comment (UniversalSuperBox) [reply qml/Keyboard.qml:388]: Please add these to the comment so future code travelers know if the bug has been fixed.

File: qml/Keyboard.qml
Comment (kugiigi) [reply qml/Keyboard.qml:603]: That's actually what I tried to avoid when I changed the vertical movement from a separate `if` to  `else if`. However, what you are encountering is mostly an effect of removing `input_method.surrounding*` from the criteria for the horizontal movements. I will try to find if there's a way to at least lessen this behavior.

Speaking of, I decided to make the cursor mover work on QtWebEngine in exchange of some trade offs such as this. I think the benefit is greater than the loss.

Comment (UniversalSuperBox) [reply qml/Keyboard.qml:603]: Hmm, okay. If you aren't able to find another way, our users will let us know if the trade-off was worth it. :)

File: qml/Keyboard.qml
Comment (kugiigi) [reply qml/Keyboard.qml:583]: I've added the link to the bug in the latest commit.

File: src/view/abstracttexteditor.cpp
Comment (kugiigi) [reply src/view/abstracttexteditor.cpp]: I've deleted the comment in the latest commit.

File: src/view/abstracttexteditor.cpp
Comment (kugiigi) [reply src/view/abstracttexteditor.cpp]: I've deleted the comment in the latest commit.

File: qml/Keyboard.qml
Comment (kugiigi) [reply qml/Keyboard.qml:337]: I've added texts in the latest commit.

File: qml/Keyboard.qml
Comment (kugiigi) [reply qml/Keyboard.qml:603]: I decided to revert my changes as it seems that it makes it a bit worse. The issue can still occur though and I can't find a way to completely avoid it.

Comment (UniversalSuperBox) [reply qml/Keyboard.qml:603]: It might need a more complex algorithm to decide if the user is trying to scroll horizontally or vertically. It's unlikely they'd do both at the same time, right?

File: qml/ActionsToolbar.qml
Comment (UniversalSuperBox) [inline qml/ActionsToolbar.qml:1]: Something new and interesting... When using this under a right-to-left language, the cursor moves opposite to your finger. I don't know how to write or read a right-to-left language, but I suspect this is not what they'd want.

File: qml/Keyboard.qml
Comment (kugiigi) [reply qml/Keyboard.qml:603]: Do you mean the code or the user? I think the answer is yes for both, it's possible that the user wants some kind of a diagonal movement and code does allow that at the moment.
The problem I think is that horizontal movement is more sensitive than vertical so when you swipe upwards or downwards, most of the time that also involves movements in the x axis.

File: qml/ActionsToolbar.qml
Comment (kugiigi) [reply qml/ActionsToolbar.qml:1]: I actually noticed this too but I haven't tried to find a fix yet since it's an existing bug. Let me check in the C++ part if we can find something we can check to invert the horizontal directions.

File: qml/Keyboard.qml
Comment (kugiigi) [reply qml/Keyboard.qml:603]: I tried some other approach but they all failed. I can't think of any other way. The property `input_method.surrounding*` is really needed to avoid this issue.

File: qml/ActionsToolbar.qml
Comment (kugiigi) [reply qml/ActionsToolbar.qml:1]: Sadly, I wasn't able to find anything that can help with this from the C++ part and QML part. To be fair, the same behavior happens in Notepad++ on Windows 10 when setting to RTL 😀 
Perhaps this is something we can just create a separate issue?

Comment (UniversalSuperBox) [reply qml/ActionsToolbar.qml:1]: Interesting. Sometimes it works, sometimes it doesn't. The cursor movement seems broken in general with RTL languages... but only with on-screen keyboard. An attached hardware keyboard can move the cursor normally.

File: qml/Keyboard.qml
Comment (UniversalSuperBox) [reply qml/Keyboard.qml:603]: Hmm, okay. It's not worse than before, in any case. I think we can accept this.

File: qml/ActionsToolbar.qml
Comment (kugiigi) [reply qml/ActionsToolbar.qml:1]: I've tried different scenarios using Arabic. All in edge channel though.
The same behavior happened when I used the OSK and a bluetooth KB in Jotit app.
I tried again in QML Creator and the same thing happened - cursor movements are reversed.
But then I tried in messaging app, same thing happened until I tapped on the space bar and the text went from left to right. I figured this happened because I have word suggestions enabled.
So I experimented further, and it seems that text fields can only be LTR or RTL and I guess it cannot be changed anymore. I had a scenario when I typed Arabic first so the text are on the right then I typed English texts and the cursor movement are now reversed in the English texts.

So my conclusion is that the behavior depends on the direction of the text field (RTL/LTR) and the direction of the actual texts. If the field is in LTR, you'll get reverse cursor movement on RTL languages and when the field is RTL, the LTR languages will get the reverse movement.
With that, I guess it'll be not much of a problem for RTL users since they most likely start with RTL in text fields.

That was a long post and hopefully helpful 😀

**Comment by UniversalSuperBox** [comment]
Unfortunately I think we'll need to wait to merge this until after OTA-10 is released so it can have more testing in the devel/rc channels. This is our fault, the release date has been a little up in the air as we've been working through issues on the edge channel. So, let's keep looking at and working through this up until the merge window opens again.

**Comment by kugiigi** [comment]
Yeah I figured that would happen and I think it's actually better to do it that way.
This is quite a change and needs testing and perhaps someone else can improve it before landing.
Or maybe I can include my bottom edge gesture :stuck_out_tongue_closed_eyes: 

Or should we land this on edge first to get more inputs and opinions?

**Comment by UniversalSuperBox** [comment]
Nah, no need to bring it in on edge. It'll be fine on devel.

**Comment by myii** [comment]
I've been a very long-term user of these advanced text functions (since November 2018 or thereabouts), perhaps I've used it even more than @kugiigi himself!  I'm planning to rebuild again at some point, using this PR as the basis.  I'll post any feedback that I've got.

**Comment by myii** [comment]
Rebuilt again and it is looking and feeling more polished than ever, great job @kugiigi!

One thing I've noticed for now:

1. When (left/backwards) scrolling across multiple lines of text, when it gets to the next line up, it has a habit of jumping back to the line it was already on -- you have to almost swipe diagonally upwards to keep it moving in the right direction.  It applies (but not so pronounced) when swiping right/forwards as well.  It could be the natural reach of the thumb when swiping but then that's always going to catch people out, so the interface should consider that.

**Comment by UniversalSuperBox** [comment]
@myii, do you mean when releasing the held point? I'm finding that as well. When I release my thumb after a leftward swipe, the cursor moves down. Likely because of anatomy and stuff.

@kugiigi, this seems to be something we won't get around. What feels like a straight swipe left really isn't. Here's me, holding my Nexus 5 with only my right hand, doing a swipe from right to left. It really felt straight.

image

**Comment by myii** [comment]
@UniversalSuperBox My workaround (to get around this anatomical issue) is to keep my swipe motion along the very bottom edge.  That avoids scrolling up or down a line by mistake.  Not sure how this can be used to reach a solution, though.

**Comment by kugiigi** [comment]
I am not sure if I really understand your observations but I think those behaviors are not impacted by my changes and perhaps worth having a separate issue to maybe improve in the future.

As for my changes, the real known behavioral change is that horizontal swipes will now move the cursor upwards or downwards when reaching the end of a line. This is just to make cursor movements work in QtWebEngine and should be changed again in the future.

**Comment by UniversalSuperBox** [comment]
The thing is, with the current design people usually are near the bottom of the screen and make an effort not to move off the screen, so there's little vertical movement of the cursor. With this change it's possible to start a swipe further from the bottom of the screen and cause the cursor to jump between lines.

**Comment by kugiigi** [comment]
Sorry but I don't see how the changes affect this specific use case. The trigger is still the same which is the space key. I haven't included yet the bottom swipe gesture 😁 
Maybe you are specifically pertaining to selection mode? 

And personally, I don't encounter this issue as often to the point that it's counter-intuitive and annoying. But perhaps I have smaller fingers and might have used the feature for so long that I kinda mastered it already 😄  

In any case, suggestions are welcome to improve the cursor mover. On top of my head, maybe we can lower the sensitivity of vertical movements. At the moment, it is already lower the horizontal but making it lower might lessen the accidental vertical movements.

**Comment by UniversalSuperBox** [comment]
To summarize what we've found, your thumb curves down as you move left (assuming you're right handed). That's not unexpected, with the thumb's range of motion. This occurs more if you start your gesture further away from the bottom of the screen, which you can do once you've activated cursor moving mode.

Maybe making the vertical movement less sensitive would help, but I think the problem will still happen at different screen sizes. It's hard to draw straight lines.

Instead, maybe it'd make more sense to have distinct "move left/right" and "move up/down" areas (like old touchpads), or to find out how Flickables reject off-axis movement and steal that. As it is, anatomy makes this difficult to interact with.

**Review by UniversalSuperBox — APPROVED** [review]
Okay, I think we've done enough deliberation. Let's get some wider feedback.

**Comment by Fuseteam** [comment]
yay

**Comment by giiba** [comment]
Just want to say this is awesome!

Best text selection I've experienced on a mobile device, and it is already becoming second nature. Thanks kugiigi.

**Comment by ziggutas** [comment]
Nexus 7 (flo) and BQ E4.5 (krillin); both on rc (2019-W41/2)
This is a great feature and works well on these two devices. I struggled to select text before, to the point where I wouldn't bother trying anymore! Great job!

**Pull request merged**
